client/client-multi/client-objective-c/Kaa/DefaultChannelManager.m

```diff
@@ -0,0 +1,600 @@
+                if (self.failoverManager) {
+                    [self.failoverManager onServerChanged:newServer];
+                } else {
+                    DDLogWarn(@"%@ Failover manager is nil", TAG);
+                }
+            }
+        }
+    }
+}
+
+- (void)onServerFailed:(id<TransportConnectionInfo>)server {
+    @synchronized(self) {
+        if (self.isShutdown) {
+            DDLogWarn(@"%@ Can't process server failure. Channel manager is down", TAG);
+            return;
+        }
+        
+        if ([server serverType] == SERVER_BOOTSTRAP) {
+            id<TransportConnectionInfo> nextConnectionInfo = [self getNextBootstrapServer:server];
+            if (nextConnectionInfo) {
+                DDLogVerbose(@"%@ Using next bootstrap server", TAG);
+                FailoverDecision *decision = [self.failoverManager onFailover:FAILOVER_STATUS_CURRENT_BOOTSTRAP_SERVER_NA];
+                switch (decision.failoverAction) {
+                    case FAILOVER_ACTION_NOOP:
+                        DDLogWarn(@"%@ No operation is performed according to failover strategy decision", TAG);
+                        break;
+                    case FAILOVER_ACTION_RETRY:
+                    {
+                        NSInteger period = [decision retryPeriod];
+                        DDLogWarn(@"%@ Reconnect to current bootstrap server will be made in %li ms", TAG, (long)period);
+                        __weak typeof(self)weakSelf = self;
+                        dispatch_time_t delay = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(period * NSEC_PER_MSEC));
+                        dispatch_after(delay, [self.executorContext getSheduledExecutor], ^{
+                            [weakSelf onTransportConnectionInfoUpdated:server];
+                        });
+                    }
+                        break;
+                    case FAILOVER_ACTION_USE_NEXT_BOOTSTRAP:
+                    {
+                        NSInteger period = [decision retryPeriod];
+                        DDLogWarn(@"%@ Connection to next bootstrap server will be made in %li ms", TAG, (long)period);
+                        __weak typeof(self)weakSelf = self;
+                        dispatch_time_t delay = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(period * NSEC_PER_MSEC));
+                        dispatch_after(delay, [self.executorContext getSheduledExecutor], ^{
+                            [weakSelf onTransportConnectionInfoUpdated:nextConnectionInfo];
+                        });
+                    }
+                        break;
+                    case FAILOVER_ACTION_STOP_APP:
+                        DDLogWarn(@"%@ Stopping application according to failover strategy decision!", TAG);
+                        exit(EXIT_FAILURE);
+                        //TODO review how to exit application
+                        break;
+                    default:
+                        break;
+                }
+            } else {
```

Comment:
According to the getNextBootstrapServer() implementation, this branch is dead code,